File: computing/interview/design-patterns/behavioral/observer/java/ObserverDemo.java
package interview.designpatterns.behavioral.observer;

import java.util.ArrayList;
import java.util.List;

interface WeatherObserver {
    void update(double temperature);
}

final class WeatherStation {
    private final List<WeatherObserver> observers = new ArrayList<>();
    private double temperature;

    void addObserver(WeatherObserver observer) {
        observers.add(observer);
    }

    void setTemperature(double temperature) {
        this.temperature = temperature;
        notifyObservers();
    }

    private void notifyObservers() {
        for (WeatherObserver observer : observers) {
            observer.update(temperature);
        }
    }
}

final class DisplayPanel implements WeatherObserver {
    private final String name;

    DisplayPanel(String name) {
        this.name = name;
    }

    @Override
    public void update(double temperature) {
        System.out.printf("%s temperature updated: %.1f°C%n", name, temperature);
    }
}

public final class ObserverDemo {
    private ObserverDemo() {}

    public static void main(String[] args) {
        WeatherStation station = new WeatherStation();
        station.addObserver(new DisplayPanel("Lobby"));
        station.addObserver(new DisplayPanel("Server Room"));

        station.setTemperature(22.5);
        station.setTemperature(24.0);
    }
}

File: computing/interview/design-patterns/behavioral/observer/csharp/ObserverDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Behavioral.Observer
{
    public interface IWeatherObserver
    {
        void Update(double temperature);
    }

    public sealed class WeatherStation
    {
        private readonly List<IWeatherObserver> _observers = new();
        private double _temperature;

        public void AddObserver(IWeatherObserver observer)
        {
            _observers.Add(observer);
        }

        public void SetTemperature(double temperature)
        {
            _temperature = temperature;
            Notify();
        }

        private void Notify()
        {
            foreach (var observer in _observers)
            {
                observer.Update(_temperature);
            }
        }
    }

    public sealed class DisplayPanel : IWeatherObserver
    {
        private readonly string _name;

        public DisplayPanel(string name)
        {
            _name = name;
        }

        public void Update(double temperature)
        {
            Console.WriteLine($"{_name} temperature updated: {temperature}°C");
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var station = new WeatherStation();
            station.AddObserver(new DisplayPanel("Lobby"));
            station.AddObserver(new DisplayPanel("Server Room"));

            station.SetTemperature(22.5);
            station.SetTemperature(24.0);
        }
    }
}
